src/sentry/snuba/outcomes.py

```diff
@@ -0,0 +1,341 @@
+from abc import ABC, abstractmethod
+from django.http import QueryDict
+from sentry.utils.snuba import raw_query
+from .dataset import Dataset
+from sentry_relay import DataCategory
+
+from sentry.snuba.sessions_v2 import (
+    get_constrained_date_range,
+    massage_sessions_result,
+    SimpleGroupBy,
+    InvalidField,
+)
+from sentry.utils.outcomes import Outcome
+from datetime import datetime
+from sentry.search.utils import InvalidQuery
+
+"""
+The new Outcomes API defines a "metrics"-like interface which is can be used in
+a similar way to "sessions" and "discover"
+# "Metrics"
+We have 2 "metrics" that we can query:
+
+- `quantity` (counter): The relevant stat based on category:
+        standard event: # of outcomes
+        session: number of events within the session
+        attachment: quantity in bytes of attachments
+- `times_seen` (counter): The number of outcomes that occurred --
+        differs from quantity in that:
+            it would count # of attachments, and sessions
+            if we ever aggregate multiple events into one outcome,
+            this would be used to get the number of outcomes
+
+# Tags / Grouping
+
+The Outcomes data can be grouped / filtered
+by these fields
+- `project`
+- `outcome`
+- `reason`
+- `category`
+"""
+
+ResultSet = List[Dict[str, Any]]
+Condition = Tuple[str, str, List[Any]]
+
+
+class Field(ABC):
+    @abstractmethod
+    def get_snuba_columns(self, raw_groupby: Optional[Sequence[str]] = None) -> List[str]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def extract_from_row(
+        self, row: Optional[Mapping[str, Any]], group: Optional[Mapping[str, Any]] = None
+    ) -> int:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        raise NotImplementedError()
+
+
+class QuantityField(Field):
+    def get_snuba_columns(self, raw_groupby: Optional[Sequence[str]] = None) -> List[str]:
+        return ["quantity"]
+
+    def extract_from_row(
+        self, row: Optional[Mapping[str, Any]], group: Optional[Mapping[str, Any]] = None
+    ) -> int:
+        if row is None:
+            return 0
+        return int(row["quantity"])
+
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        return ["sum", "quantity", "quantity"]
+
+
+class TimesSeenField(Field):
+    def get_snuba_columns(self, raw_groupby: Optional[Sequence[str]] = None) -> List[str]:
+        return ["times_seen"]
+
+    def extract_from_row(
+        self, row: Optional[Mapping[str, Any]], group: Optional[Mapping[str, Any]] = None
+    ) -> int:
+        if row is None:
+            return 0
+        return int(row["times_seen"])
+
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        if dataset == Dataset.Outcomes:
+            return ["sum", "times_seen", "times_seen"]
+        else:
+            # RawOutcomes doesnt have times_seen, do a count instead
+            return ["count()", "", "times_seen"]
```

Comment:
What's the deal with the three aggregation values? Do the three positions have any significance? Is it documented anywhere?

A dataclass with three fields could improve on `List[str]`, but probably not worth the extra lines without 3.8's dataclass feature.

If the value is constrained to always have the same length, I think a tuple matches Python conventions better than a list. You can do `Tuple[str, str, str]` as the type hint, right?

Reply:
Just FYI, dataclasses are available as a third party library in Python <3.8.

> You can do Tuple[str, str, str] as the type hint, right?

Just a note that since the return value isn't a Tuple mypy wouldn't like that type hinting.

Reply:
@RyanSkonnord that piece is documented in `SnubaQueryParams` in `utils/snuba.py` : 
> `aggregations` a list of (aggregation_function, column, alias) tuples to be passed to the query.

Definitely agree this could be more obvious with types / NamedTuples / DataClasses. Tried to use some NamedTuples for some of the types but we JSON serialize NamedTuples as JSON objects instead of an array, so didn't work out there. For now, i've converted this return type to Tuple[str,str,str] as it is more descriptive.